Remove SampleControlPoint and DifficultyControlPoint from HitObject

OliBomby opened the pull request:
The purpose is to clean up the `HitObject` class by not allowing any redundant information to be stored by the `HitObject`.
- The slider velocity and tick generating information has been moved to `Slider` and got a separate interface for access.
- Legacy information is now property of the `HitObject`.
- Sample control point information gets applied to the `Samples` directly,  making the sample control point information completely obsolete.

fixes #22171 #22132 #15618 #19483

OliBomby committed:
made the SampleControlPoint and DifficultyControlPoint obsolete

OliBomby committed:
Add LegacyContext

OliBomby committed:
remove ApplyLegacyInfo method

OliBomby committed:
Fix incorrect xmldoc

OliBomby committed:
remove unused directives

OliBomby committed:
remove hitobject SampleControlPoint usage from LegacyBeatmapEncoder

OliBomby committed:
Fix file header notice

OliBomby committed:
remove other usages of hitobject SampleControlPoint

OliBomby committed:
Fixed DifficultyPointPiece

OliBomby committed:
add missing }

OliBomby committed:
fix implementations of IHasSliderVelocity

OliBomby committed:
clean code

OliBomby committed:
clean up code stuff

bdach commented:
Test failures look very not promising at all.

bdach commented on osu.Game/Context/IContext.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace osu.Game.Context;
+
+public interface IContext
```
All of this `IContext` machinery needs better naming and xmldoc at the least, or preferably not exist at all. I have no idea what anything is supposed to be here. It looks to be a generic "attach anything to everything" mechanism for whose existence I see no real reason. Is there one? Is this taking some particular far-off direction that I'm failing to notice?

My questions would be in order:

- Does this PR need to touch difficulty points at all, and if so, then why? I can understand wanting get rid of sample points, because of https://github.com/ppy/osu/issues/15618#issuecomment-1224120633 and https://github.com/ppy/osu/issues/22171, but I don't see much point or reasoning in doing this for difficulty points.
- As far as I can see, right now this only serves the purpose of attaching the "legacy bpm multiplier" stuff. Could that issue be solved in a different way that does not require this baroqueness?

OliBomby replied:
- Touching the difficulty points came from the same idea as removing sample control points from `HitObject', namely "Why have any 'control point' stored in `HitObject` if it only functions as a container for some properties of the `HitObject`?". And difficulty control points were particularly weird, they store a slider velocity for all hit objects even though that's only relevant for sliders, and the legacy information just seems tacked on just because it was possible. Since I'm already changing the whole data flow, might as well fix that too.
- I gotta admit the `IContext` stuff is quite experimental. Removing it would result in directly adding the legacy bpm multiplier stuff as a property of `HitObject` and I'd need some other way of passing the information about `GenerateTicks` from the legacy decoder to the `Slider` class. The `IContext` system made it possible to attach this information to a `HitObject` without cluttering it with legacy properties.

bdach replied:
Just to clarify: don't misunderstand me here, I'm completely fine with `IHasSliderVelocity` et al. I'm much less fine with the API of "attach anything you like here", because if we open that Pandora box, there's no telling what ruleset authors are going to go and do with it.

All I'm saying is that I'd want to have a more constrained interface here. And what I'd be okay is probably a middle ground of what you have here and what we have in `master` - namely, the interface still only applies to things where slider velocity makes sense, but the member exposed by the interface would be essentially what is currently the difficulty control point.

OliBomby replied:
I've removed `IContext` and added an interface `IHasGenerateTicks` to allow storing the `GenerateTicks` property in `Slider` and I added the `LegacyBpmMultiplier` to `HitObject`.

OliBomby committed:
Invoke ApplyDefaultsToSelf

OliBomby committed:
Remove IContext & add IHasGenerateTicks

OliBomby committed:
removing using

OliBomby committed:
Update SamplePointPiece.cs

OliBomby committed:
remove all non-test usage of SampleControlPoint

OliBomby committed:
remove all test usages of SampleControlPoint

OliBomby committed:
fix invalidoperation exception

OliBomby committed:
fixed sample control point applying

OliBomby committed:
fix usings

OliBomby committed:
remove all regular usage of DifficultyControlPoint

OliBomby committed:
removed all usage of hitobject's DifficultyControlPoint

OliBomby committed:
remove bank default value

OliBomby committed:
fix legacy parser incorrectly assigning sample info for sliders

OliBomby committed:
fix sample of drumroll ticks being bankless

OliBomby committed:
Update CheckMutedObjectsTest.cs

OliBomby committed:
fix null sample

OliBomby committed:
fix ticks not being generated by default

OliBomby committed:
Update convert-samples-expected-conversion.json

OliBomby committed:
fix 0 velocity juicestream

OliBomby committed:
fix legacy encoder writing sample info while not writing node samples

OliBomby committed:
pass new hitobject properties through beatmap converters

OliBomby committed:
add min and max value to SliderVelocity

OliBomby committed:
Merge remote-tracking branch 'upstream/master' into sample-control-points

OliBomby committed:
Make sure the first object you place has bank and volume

OliBomby committed:
dont assign custom sample bank and volume to hitobjects in non-mania gamemodes

this makes it easier to edit hitsounds in the stable editor after export because the sample control point effects wont get overwritten by the properties of the hitobject

OliBomby committed:
add sample info to Banana and SpinnerBonusTick

OliBomby committed:
dont encode custom sample bank for objects without legacy samples

OliBomby committed:
remove all redundancies from encoded control points

peppy committed:
Merge branch 'master' into sample-control-points

peppy commented:
Before dropping a review, huge thanks for taking on this effort. It's reassuring to have you both willing to understand and dive deep into the low-level structure of things surrounding hitobjects (which still have some work to get into a less-complex state).

peppy commented on osu.Game/Rulesets/Objects/Types/IHasGenerateTicks.cs:
```diff
@@ -0,0 +1,17 @@
+    public interface IHasGenerateTicks
+    {
+        /// <summary>
+        /// Whether or not slider ticks should be generated at this control point.
```
```suggestion
        /// Whether or not slider ticks should be generated by this object.
```

peppy commented on osu.Game/Rulesets/Objects/Types/IHasGenerateTicks.cs:
```diff
@@ -0,0 +1,17 @@
+namespace osu.Game.Rulesets.Objects.Types
+{
+    /// <summary>
+    /// A type of <see cref="HitObject"/> which may or may not generate ticks.
```
```suggestion
    /// A type of <see cref="HitObject"/> which explicitly specifies whether it should generate ticks.
```

peppy committed:
Fix non-block namespace usage

OliBomby commented on osu.Game.Rulesets.Catch/Edit/Blueprints/Components/EditablePath.cs:
```diff
@@ -103,7 +103,7 @@ public void UpdateHitObjectFromPath(JuiceStream hitObject)
             // The value is clamped here by the bindable min and max values.
             // In case the required velocity is too large, the path is not preserved.
-            svBindable.Value = Math.Ceiling(requiredVelocity / svToVelocityFactor);
+            svBindable.Value = requiredVelocity == 0 ? 1 : Math.Ceiling(requiredVelocity / svToVelocityFactor);
```
While fixing tests, I ran into the case where a `JuiceStream` has control points all on `(0,0)` which resulted in a 0 length path which set the slider velocity to 0. This 0 velocity then resulted in a bunch of divide by zero problems such as `NaN` time nested hit objects. Simply not allowing the velocity to reach 0 fixed this.

I later added the `MinimumValue` thing to the slider velocity bindable so maybe this change isn't necessary anymore.

OliBomby committed:
Fix comment 1 osu.Game/Rulesets/Objects/Types/IHasGenerateTicks.cs

Co-authored-by: Dean Herbert <pe@ppy.sh>

OliBomby committed:
Fix comment 2 osu.Game/Rulesets/Objects/Types/IHasGenerateTicks.cs

Co-authored-by: Dean Herbert <pe@ppy.sh>

OliBomby commented on osu.Game.Rulesets.Taiko/Objects/DrumRoll.cs:
```diff
@@ -81,7 +97,8 @@ private void createTicks(CancellationToken cancellationToken)
                     StartTime = t,
-                    IsStrong = IsStrong
+                    IsStrong = IsStrong,
+                    Samples = Samples.Where(s => s.Name == HitSampleInfo.HIT_FINISH).ToList()
```
What do you mean by this? Should i make a test case for this line of code?

peppy replied:
This was a self-note, I just want to make sure the taiko stuff is correct as i'm not 100% we have coverage of it. Same applies to the other comment below.

OliBomby committed:
remove redundant zero check in sv calculation

OliBomby committed:
remove constructor argument from Banana

peppy committed:
Ensure `static` banana samples are not mutated

peppy committed:
Move pragma disable to top of `LegacyBeatmapDecoder`

Makes more sense as it's used multiple times in the class.

peppy committed:
Refactor `applyLegacyInfoAndDefaults` for legibility

peppy committed:
Split out default and sample application

peppy committed:
Move `struct` to bottom of file

peppy commented:
I'm going to get these changes in. Pretty happy with how they are looking. I am sure we'll find a few weird issues post-merge but would rather keep moving forward.

For reference, here's benchmark results before and after:

`master`:

|                  Method |     Mean |     Error |    StdDev |     Gen0 |     Gen1 | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 6.744 ms | 0.1282 ms | 0.1199 ms | 343.7500 | 171.8750 |   3.78 MB |



|     Method | Count | WithBindableAccess |         Mean |       Error |      StdDev |     Gen0 |    Gen1 | Allocated |
|----------- |------ |------------------- |-------------:|------------:|------------:|---------:|--------:|----------:|
|  OsuCircle |     1 |              False |     178.9 ns |     1.47 ns |     1.30 ns |   0.0632 |  0.0002 |     728 B |
|   TaikoHit |     1 |              False |     316.8 ns |     2.97 ns |     2.63 ns |   0.0949 |  0.0005 |    1096 B |
| CatchFruit |     1 |              False |     126.7 ns |     1.46 ns |     1.29 ns |   0.0520 |       - |     600 B |
|  ManiaNote |     1 |              False |     114.8 ns |     1.62 ns |     1.51 ns |   0.0381 |       - |     440 B |
|  OsuCircle |     1 |               True |     308.5 ns |     6.16 ns |     5.46 ns |   0.1168 |  0.0010 |    1344 B |
|   TaikoHit |     1 |               True |     320.5 ns |     2.47 ns |     2.31 ns |   0.0949 |  0.0005 |    1096 B |
| CatchFruit |     1 |               True |     284.2 ns |     2.80 ns |     2.48 ns |   0.1192 |  0.0010 |    1376 B |
|  ManiaNote |     1 |               True |     130.1 ns |     2.58 ns |     2.87 ns |   0.0458 |       - |     528 B |

this PR:

|                  Method |     Mean |     Error |    StdDev |     Gen0 |     Gen1 | Allocated |
|------------------------ |---------:|----------:|----------:|---------:|---------:|----------:|
| BenchmarkBundledBeatmap | 5.766 ms | 0.0795 ms | 0.0744 ms | 328.1250 | 164.0625 |   3.66 MB |



|     Method | Count | WithBindableAccess |         Mean |       Error |      StdDev |     Gen0 |    Gen1 | Allocated |
|----------- |------ |------------------- |-------------:|------------:|------------:|---------:|--------:|----------:|
|  OsuCircle |     1 |              False |     174.9 ns |     1.88 ns |     1.76 ns |   0.0639 |  0.0002 |     736 B |
|   TaikoHit |     1 |              False |     307.9 ns |     2.44 ns |     2.28 ns |   0.0949 |  0.0005 |    1096 B |
| CatchFruit |     1 |              False |     125.7 ns |     1.57 ns |     1.47 ns |   0.0527 |  0.0002 |     608 B |
|  ManiaNote |     1 |              False |     109.5 ns |     0.43 ns |     0.40 ns |   0.0381 |  0.0001 |     440 B |
|  OsuCircle |     1 |               True |     288.6 ns |     2.65 ns |     2.48 ns |   0.1173 |  0.0010 |    1352 B |
|   TaikoHit |     1 |               True |     329.8 ns |     4.86 ns |     4.31 ns |   0.0949 |  0.0005 |    1096 B |
| CatchFruit |     1 |               True |     284.9 ns |     2.94 ns |     2.61 ns |   0.1202 |  0.0010 |    1384 B |
|  ManiaNote |     1 |               True |     125.5 ns |     2.15 ns |     2.01 ns |   0.0458 |       - |     528 B |

OliBomby commented:
You can probably close these issues as of the merging of this PR: https://github.com/ppy/osu/issues/22132 https://github.com/ppy/osu/issues/15618 https://github.com/ppy/osu/issues/19483
